### PR TITLE
HIVE-21599: Parquet predicate pushdown on partition columns may cause wrong result if files contain partition columns

### PR DESCRIPTION
**Pull request by soumyakanti3578**
### What changes were proposed in this pull request?
Partition columns are getting removed from Parquet metadata (schema).

### Why are the changes needed?
When a Parquet data file contains partition columns, and the query filters on those partition columns, we can get wrong results. By removing the partition columns from the schema, we avoid creating Filter predicates on those columns.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
`mvn test  -Dtest=TestMiniLlapLocalCliDriver -Dtest.output.overwrite=true -Dqfile=parquet_partition_col.q`
This test returns correct results.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -4272,6 +4273,21 @@ public static void addTableSchemaToConf(Configuration conf,
...
   }
+  public static void setPartitionColumnsToConf(Configuration conf, TableScanOperator tableScanOp) {
+    TableScanDesc scanDesc = tableScanOp.getConf();
+    if (scanDesc != null && scanDesc.getTableMetadata() != null) {
+      List<String> partitionColsList = scanDesc.getTableMetadata().getPartColNames();
+      if (!partitionColsList.isEmpty()) {
+        conf.set(IOConstants.PARTITION_COLUMNS, String.join(",", partitionColsList));
+      }
+    } else {
+      LOG.info(IOConstants.PARTITION_COLUMNS + " not available");
```
Comment (amansinha100) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4284]: Is this else condition placed correctly ?  Seems like the INFO message  for partition columns not available  should be logged when partitionColsList.isEmpty() returns true.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java [hunk 2]
```diff
@@ -209,6 +214,22 @@ public FilterCompat.Filter setFilter(final JobConf conf, MessageType schema) {
...
   }
 
+  private MessageType getSchemaWithoutPartitionColumns(JobConf conf, MessageType schema) {
+    String partCols = conf.get(IOConstants.PARTITION_COLUMNS);
+    if (partCols != null && partCols.length() > 0) {
+      Set<String> partitionColumns = new HashSet<>(Arrays.asList(partCols.split(",")));
+      List<Type> newFields = new ArrayList<>();
+
+      for (Type field: schema.getFields()) {
+        if(!partitionColumns.contains(field.getName())) {
```
Comment (amansinha100) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java:224]: Could you confirm if this comparison works in a case-insensitive way ? i.e if the partition columns is defined as 'part_col' and the query has predicate on 'PART_COL'.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 3]
```diff
@@ -4272,6 +4273,21 @@ public static void addTableSchemaToConf(Configuration conf,
       LOG.info("schema.evolution.columns and schema.evolution.columns.types not available");
     }
   }
+  public static void setPartitionColumnsToConf(Configuration conf, TableScanOperator tableScanOp) {
```
Comment (amansinha100) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4276]: nit: pls add a comment for this method.  Also, the unset method is named 'InConf' whereas this one has 'ToConf'.  Can both set and unset method names use the same suffix ?

**Comment by amansinha100** [comment]
A note about the commit message: the summary says 'remove predicate ..' which is not true any more with this patch. So best to reword it.  Also remove reference to virtual columns since the patch is not making any changes for that.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 4]
```diff
@@ -4272,6 +4273,21 @@ public static void addTableSchemaToConf(Configuration conf,
...
+      List<String> partitionColsList = scanDesc.getTableMetadata().getPartColNames();
+      if (!partitionColsList.isEmpty()) {
+        conf.set(IOConstants.PARTITION_COLUMNS, String.join(",", partitionColsList));
+      }
+    } else {
+      LOG.info(IOConstants.PARTITION_COLUMNS + " not available");
+    }
+  }
+
+  public static void unsetPartitionColumnsInConf(Configuration conf) {
```
Comment (asolimando) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4288]: Public methods should have javadoc added.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java [hunk 5]
```diff
@@ -209,6 +214,22 @@ public FilterCompat.Filter setFilter(final JobConf conf, MessageType schema) {
     }
   }
 
+  private MessageType getSchemaWithoutPartitionColumns(JobConf conf, MessageType schema) {
+    String partCols = conf.get(IOConstants.PARTITION_COLUMNS);
+    if (partCols != null && partCols.length() > 0) {
```
Comment (asolimando) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java:219]:
```suggestion
    if (partCols != null && !partCols.isEmpty()) {
```
Nit: maybe using isEmpty here can help?

**Review by asolimando — CHANGES_REQUESTED** [review]
LGTM, only please add the missing javadoc on newly introduced public methods

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 1 above)
Comment (soumyakanti3578) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4284]: Thanks for catching this. I will move it to the correct place.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java (hunk 2 above)
Comment (soumyakanti3578) [reply ql/src/java/org/apache/hadoop/hive/ql/io/parquet/ParquetRecordReaderBase.java:224]: I tested with `PCOL` and  \`pCol\`, and it's working fine. I have added more queries to the q file.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 3 above)
Comment (soumyakanti3578) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4276]: Setting both to InConf.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 4 above)
Comment (soumyakanti3578) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:4288]: Thanks, I will add the docs!

**Review by amansinha100 — APPROVED** [review]
+1

File: ql/src/test/queries/clientpositive/parquet_partition_col.q [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ select * from test;
+
+ explain select * from test where pcol=100 and intcol=2;
+ select * from test where pcol=100 and intcol=2;
+
+ explain select * from test where PCOL=200 and intcol=3;
+ select * from test where PCOL=200 and intcol=3;
+
+ explain select * from test where `pCol`=300 and intcol=5;
+ select * from test where `pCol`=300 and intcol=5;
```
Comment (asolimando) [inline ql/src/test/queries/clientpositive/parquet_partition_col.q:20]: Nit: text files should end with a newline to be well formed

**Review by asolimando — APPROVED** [review]
+1, for the future avoid force pushing/rebasing while the review process is still on-going because this kills the possibility to review only the delta from the previous review round

**Review by zabetak — COMMENTED** [review]
While reviewing this PR, I got the impression that the solution may be simpler and more general. I left some comments under the JIRA ticket and pushed an alternative fix [here](https://github.com/apache/hive/pull/3742/commits/7e16714b28fb3611bff1d9f35be79c400197055b).

@soumyakanti3578 @amansinha100 @asolimando let me know your thoughts.

**Comment by soumyakanti3578** [comment]
@zabetak LGTM. Thanks for cleaning this up!

**Review by amansinha100 — APPROVED** [review]
Updated changes look good to me.  +1 .

**Pull request closed without merging**